File: LegendDataEditor/LegendDataEditor/Sqlite/Table/SQLiteMonsterFrame.cpp
#include "SQLiteMonsterFrame.h"
#include "Utility.h"
#include "SQLiteDataReader.h"
#include "SQLite.h"

string MonsterFrameData::COL_ID = "MonsterID";
string MonsterFrameData::COL_LABEL = "Label";
string MonsterFrameData::COL_DIRECTION = "Direction";
string MonsterFrameData::COL_ACTION = "Action";
string MonsterFrameData::COL_FRAME_COUNT = "FrameCount";
string MonsterFrameData::COL_POSX = "PosX";
string MonsterFrameData::COL_POSY = "PosY";

void SQLiteMonsterFrame::query(int monsterID, txVector<MonsterFrameData>& dataList)
{
	string conditionString;
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_ID, monsterID, "");
	string queryStr = "SELECT * FROM " + mTableName + " WHERE " + conditionString;
	parseReader(mSQLite->executeQuery(queryStr), dataList);
}
void SQLiteMonsterFrame::query(int monsterID, int direction, txVector<MonsterFrameData>& dataList)
{
	string conditionString;
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_ID, monsterID, " and ");
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_DIRECTION, direction, "");
	string queryStr = "SELECT * FROM " + mTableName + " WHERE " + conditionString;
	parseReader(mSQLite->executeQuery(queryStr), dataList);
}
void SQLiteMonsterFrame::query(int monsterID, int direction, const string& action, txVector<MonsterFrameData>& dataList)
{
	string conditionString;
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_ID, monsterID, " and ");
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_DIRECTION, direction, " and ");
	StringUtility::appendConditionString(conditionString, MonsterFrameData::COL_ACTION, action, "");
	string queryStr = "SELECT * FROM " + mTableName + " WHERE " + conditionString;
	parseReader(mSQLite->executeQuery(queryStr), dataList);
}
bool SQLiteMonsterFrame::updateData(const MonsterFrameData& data)
{
	string updateString;
	StringUtility::appendUpdateInt(updateString, MonsterFrameData::COL_FRAME_COUNT, data.mFrameCount);
	StringUtility::appendUpdateIntArray(updateString, MonsterFrameData::COL_POSX, data.mPosX);
	StringUtility::appendUpdateIntArray(updateString, MonsterFrameData::COL_POSY, data.mPosY);
	StringUtility::removeLastComma(updateString);
	string conditionString;
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_ID, data.mID, " and ");
	StringUtility::appendConditionInt(conditionString, MonsterFrameData::COL_DIRECTION, data.mDirection, " and ");
	StringUtility::appendConditionString(conditionString, MonsterFrameData::COL_ACTION, data.mAction, "");
	return doUpdate(updateString, conditionString);
}

bool SQLiteMonsterFrame::insert(const MonsterFrameData& data)
{
	string valueString;
	data.insert(valueString);
	StringUtility::removeLastComma(valueString);
	return doInsert(valueString);
}

File: LegendEditor/Assets/Scripts/Game/DataBase/SQLiteTable/SQLiteMonsterFrame.cs
using Mono.Data.Sqlite;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class MonsterFrameData : TableData
{
	public static string COL_ID = "MonsterID";
	public static string COL_LABEL = "Label";
	public static string COL_DIRECTION = "Direction";
	public static string COL_ACTION = "Action";
	public static string COL_FRAME_COUNT = "FrameCount";
	public static string COL_POSX = "PosX";
	public static string COL_POSY = "PosY";
	public int mID;
	public string mLabel;
	public int mDirection;
	public string mAction;
	public int mFrameCount;
	public List<int> mPosX;
	public List<int> mPosY;
	public override void parse(SqliteDataReader reader)
	{
		mID = stringToInt(reader[COL_ID].ToString());
		mLabel = reader[COL_LABEL].ToString();
		mDirection = stringToInt(reader[COL_DIRECTION].ToString());
		mAction = reader[COL_ACTION].ToString();
		mFrameCount = stringToInt(reader[COL_FRAME_COUNT].ToString());
		stringToIntArray(reader[COL_POSX].ToString(), ref mPosX);
		stringToIntArray(reader[COL_POSY].ToString(), ref mPosY);
	}
}

public class SQLiteMonsterFrame : SQLiteTable
{
	public SQLiteMonsterFrame()
		:base("MonsterFrame")
	{}
	public void query(int id, out List<MonsterFrameData> dataList)
	{
		string condition = "";
		appendConditionInt(ref condition, MonsterFrameData.COL_ID, id, "");
		parseReader(doQuery(condition), out dataList);
	}
	public void query(int id, int direction, out List<MonsterFrameData> dataList)
	{
		string condition = "";
		appendConditionInt(ref condition, MonsterFrameData.COL_ID, id, " and ");
		appendConditionInt(ref condition, MonsterFrameData.COL_DIRECTION, direction, "");
		parseReader(doQuery(condition), out dataList);
	}
	public void query(int id, int direction, string action, out List<MonsterFrameData> dataList)
	{
		string condition = "";
		appendConditionInt(ref condition, MonsterFrameData.COL_ID, id, " and ");
		appendConditionInt(ref condition, MonsterFrameData.COL_DIRECTION, direction, " and ");
		appendConditionString(ref condition, MonsterFrameData.COL_ACTION, action, "");
		parseReader(doQuery(condition), out dataList);
	}
	public void updateData(MonsterFrameData data)
	{
		string updateString = "";
		appendUpdateInt(ref updateString, MonsterFrameData.COL_FRAME_COUNT, data.mFrameCount);
		appendUpdateIntArray(ref updateString, MonsterFrameData.COL_FRAME_COUNT, data.mPosX);
		appendUpdateIntArray(ref updateString, MonsterFrameData.COL_FRAME_COUNT, data.mPosY, true);
		string condition = "";
		appendConditionInt(ref condition, MonsterFrameData.COL_ID, data.mID, " and ");
		appendConditionInt(ref condition, MonsterFrameData.COL_DIRECTION, data.mDirection, " and ");
		appendConditionString(ref condition, MonsterFrameData.COL_ACTION, data.mAction, "");
		doUpdate(updateString, condition);
	}
	public void insert(MonsterFrameData data)
	{
		string valueString = "";
		appendValueInt(ref valueString, data.mID);
		appendValueString(ref valueString, data.mLabel);
		appendValueInt(ref valueString, data.mDirection);
		appendValueString(ref valueString, data.mAction);
		appendValueInt(ref valueString, data.mFrameCount);
		appendValueIntArray(ref valueString, data.mPosX);
		appendValueIntArray(ref valueString, data.mPosY, true);
		doInsert(valueString);
	}
};